embassy-stm32/src/adc/f1.rs

```diff
@@ -221,6 +238,75 @@ impl<'d, T: Instance> Adc<'d, T> {
         self.convert()
     }
 
+    pub async fn read_continuous<S, const N: usize>(
+        &mut self,
+        pin: &mut impl AdcPin<T>,
+        data: &mut [[u16; N]; 2],
+        sampler: &mut S,
+    ) where
+        S: FnMut(&[u16; N]) -> SamplerState,
+        RXDMA: RxDma<T>,
+    {
+        let rx_request = self.rxdma.request();
+        let rx_src = T::regs().dr().ptr() as *mut u16;
+
+        unsafe {
+            self.rxdma
+                .start_circular_read(rx_request, rx_src, data.as_mut_ptr(), N * 2, Default::default());
+        }
+
+        unsafe {
+            Self::set_channel_sample_time(pin.channel(), self.sample_time);
+            T::regs().cr1().modify(|reg| {
+                reg.set_scan(false);
+                reg.set_discen(false);
+            });
+            T::regs().sqr1().modify(|reg| reg.set_l(0));
+
+            T::regs().cr2().modify(|reg| {
+                reg.set_cont(true); //Goes with circular DMA
+                reg.set_exttrig(true);
+                reg.set_swstart(false);
+                reg.set_extsel(crate::pac::adc::vals::Extsel::SWSTART);
+                reg.set_dma(true);
+            });
+        }
+
+        // Configure the channel to sample
+        unsafe { T::regs().sqr3().write(|reg| reg.set_sq(0, pin.channel())) }
+
+        //Enable ADC
+        unsafe {
+            T::regs().cr2().modify(|reg| {
+                reg.set_adon(true);
+                reg.set_swstart(true);
+            });
+        }
+        let mut buf_index = 0;
+        //Loop for retrieving data
+        poll_fn(|cx| {
+            self.rxdma.set_waker(cx.waker());
+
+            if self.rxdma.is_data_ready() {
+                let sampler_state = sampler(&data[buf_index]);
+                self.rxdma.set_data_processing_done();
+
+                if sampler_state == SamplerState::Sampled {
+                    buf_index = !buf_index & 0x01; // switch the buffer index (0/1)
+                    return Poll::Pending;
+                } else {
+                    return Poll::Ready(());
+                }
+            } else {
+                return Poll::Pending;
+            }
+        })
+        .await;
```

Comment:
Phase 3, the outer part of setting waker, checking if data is ready etc. is generic, but the inner is not.